### ISSUE TITLE
Allow simultaneous parallel writes to Events

### ISSUE THREAD

**Issue by alice-i-cecile** [bevyengine/bevy#2648]
## What problem does this solve or what need does it fill?

Users cannot send multiple events (via `EventWriter`) in parallel, either between systems or within a single system when using `par_for_each` iteration.

## What solution would you like?

Change the access model of `EventWriter` to allow for multiple simultaneous append-only writes.

## What alternative(s) have you considered?

Write to events in sequence, slowly.

## Additional context

`Commands` uses a similar parallelization strategy to allow multiple systems to add new commands simultaneously.

The logic for the final ordering of these events should match the ordering of the commands, which is determined by

**Comment by Ratysz** [comment]
The ordering strategy of commands only makes sense between systems, it's not possible to apply it to inner parallelisation provided by `par_for_each` and the like; see also #1532. We can enable "simultaneous" writes, but I don't think there's a good one-size-fits-all generalised way of ordering things here, and the user should be aware that the order is essentially non-deterministic.

**Comment by alice-i-cecile** [comment]
> The ordering strategy of commands only makes sense between systems, it's not possible to apply it to inner parallelisation provided by par_for_each and the like

Yep, this ordering strategy should only apply to events that are run in parallel between systems, not within systems.

**Comment by JoJoJet** [comment]
This would be a very useful feature.

I believe that in general, a system should either have explicit ordering, or be ordering-agnostic. That is, it should not matter which order the two systems get executed in. This means turning any mutable references into read-only ones, and performing all mutation using either `Commands` or events.

This pattern works excellently for avoiding bugs due to nondeterministic scheduling, but unfortunately it doesn't improve performance since `EventWriter`s are not allowed to run in parallel with one another.

If there's any way I could contribute, let me know.

**Comment by JoJoJet** [comment]
A proposal:

* All events sent from a given `EventWriter` should have preserved ordering, and be grouped together in `EventReader::iter`. The ordering of `EventWriter` groups within the iterator is *not* guaranteed.

Here's an example.

```rust
// these run in parallel!
fn sys1(mut w: EventWriter<...>) {
    w.send("hey");
    w.send("hi");
    w.send("hello");
}
fn sys2(mut w: EventWriter<...> {
    w.send("goodbye");
    w.send("adios");
}
```
Sometimes, an `EventReader` may yield `["hey", "hi", "hello", "goodbye", "adios"]`, and sometimes it may yield `["goodbye", "adios", "hey", "hi", "hello"]`, but the groups will never be broken apart, and their ordering will never be disturbed.

This probably means that each `EventWriter` would hold a buffer, which gets flushed somehow after its system stops running. Hopefully that would be the one and only sync point to worry about (for that system).

**Comment by james7132** [comment]
Perhaps using `Parallel<T>` , introduced in #7348, could be used, though the collection and ordering of those events could be pretty nasty. It would also likely increase the storage footprint for events by a lot.

**Comment by AntiAnimeGeneral** [comment]
I personally use (https://crates.io/crates/bevy_concurrent_event) in conjunction with Renet to send network IO. 
It is similar to commands, where each system can concurrently send and read events.
However, it is still very rough and simple, and can only be used in scenarios with low latency requirements
